File: scripts/bazaar-regex-check.mjs
import { spawnSync } from "node:child_process";
import { fileURLToPath } from "node:url";
import { loadBazaarSchemas } from "./lib/load-bazaar-schemas.mjs";
const patterns = [];
function walk(item, value, path = "schema") {
  if (!value || typeof value !== "object") return;
  for (const [key, child] of Object.entries(value)) {
    if (key === "pattern" && typeof child === "string") patterns.push({ item, path: `${path}.${key}`, pattern: child });
    walk(item, child, `${path}.${key}`);
  }
}
for (const { item, schema } of await loadBazaarSchemas()) walk(item, schema);
if (!patterns.length) throw new Error("No published patterns were read; the checker did not run.");
const result = spawnSync(process.env.BAZAAR_GO_BINARY || "go", ["run", fileURLToPath(new URL("./bazaar-regex-check.go", import.meta.url))], { input: JSON.stringify(patterns), encoding: "utf8" });
if (result.error) { console.error("Go is required for this compatibility check; set BAZAAR_GO_BINARY if it is not on PATH."); process.exit(2); }
process.stdout.write(result.stdout);
process.stderr.write(result.stderr);
process.exit(result.status ?? 2);

File: scripts/bazaar-regex-check.go
// Compile the actual published patterns with Go's regex engine. JavaScript
// accepting the same pattern does not establish facilitator compatibility.
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"regexp"
)

type Pattern struct{ Item, Path, Pattern string }

func main() {
	var patterns []Pattern
	if err := json.NewDecoder(os.Stdin).Decode(&patterns); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(2)
	}
	failed := false
	for _, p := range patterns {
		if _, err := regexp.Compile(p.Pattern); err != nil {
			fmt.Fprintf(os.Stderr, "%s %s: %v\n", p.Item, p.Path, err)
			failed = true
		}
	}
	if failed {
		os.Exit(1)
	}
	fmt.Printf("%d published schema patterns compile with Go regexp.\n", len(patterns))
}
